src/realm/object_converter.cpp

```diff
@@ -253,29 +244,210 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
             }
+            // return true;
         }
     }
+    return false;
 }
 
+bool InterRealmValueConverter::copy_dictionary_in_mixed(const Dictionary& src_dict, Dictionary& dst_dict,
+                                                        bool* update_out)
+{
+    // utility functions
+
+    auto handle_set_copy = [&, this](StringData key) {
+        dst_dict.insert_collection(key, CollectionType::Set);
+        auto n_src_set = src_dict.get_set(key);
+        auto n_dst_set = dst_dict.get_set(key);
+        copy_set(*n_src_set, *n_dst_set, update_out);
+    };
+    auto handle_list_copy = [&, this](StringData key) {
+        dst_dict.insert_collection(key, CollectionType::List);
+        auto n_src_list = src_dict.get_list(key);
+        auto n_dst_list = dst_dict.get_list(key);
+        copy_list_in_mixed(*n_src_list, *n_dst_list, update_out);
+
+        // if ()
+        //     copy_list(*n_src_list, *n_dst_list, update_out);
+    };
+    auto handle_dict_copy = [&, this](StringData key) {
+        dst_dict.insert_collection(key, CollectionType::Dictionary);
+        auto n_src_dict = src_dict.get_dictionary(key);
+        auto n_dst_dict = dst_dict.get_dictionary(key);
+        copy_dictionary_in_mixed(*n_src_dict, *n_dst_dict, update_out);
+        // if () {
+        //     copy_dictionary(*n_src_dict, *n_dst_dict, update_out);
+        // }
+    };
+
+    // main algorithm
+    auto n = src_dict.size();
+    for (size_t i = 0; i < n; ++i) {
+        auto [key, any] = src_dict.get_pair(i);
+        if (any.is_type(type_List)) {
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_List);
+            handle_list_copy(key.get_string());
+        }
+        else if (any.is_type(type_Set)) {
+            // nested collections do not support nested sets, set can only be a terminal leaf
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Set);
+            handle_set_copy(key.get_string());
+        }
+        else if (any.is_type(type_Dictionary)) {
+            if (dst_dict.contains(key))
+                handle_collection_in_mixed_mismatch(any, dst_dict.get(key), type_Dictionary);
+            handle_dict_copy(key.get_string());
+        }
+        else {
+            // copy single element
+            InterRealmValueConverter::ConversionResult converted_src;
+            auto dst_obj = dst_dict.get_obj();
+            auto dst_col = dst_dict.get_col_key();
+            auto dst_mixed = dst_obj.get_any(dst_col);
+            if (cmp_src_to_dst(any, dst_mixed, &converted_src, update_out)) {
+                dst_dict.insert(key, converted_src.converted_value);
+            }
+            // return true;
+        }
+    }
+    return false;
+}
+
+void InterRealmValueConverter::handle_collection_in_mixed_mismatch(Mixed src, Mixed dst, DataType type)
```

Comment:
nit: I think "validate..." or "check..." is more appropriate for what this method does than "handle..."